src/charm.py

```diff
@@ -1,19 +1,497 @@
+            )
+            return False
+
+        return True
+
+    def get_mongos_host(self) -> str:
+        """Returns the host for mongos as a str.
+
+        The host for mongos can be either the Unix Domain Socket or an IP address depending on how
+        the client wishes to connect to mongos (inside Juju or outside).
+        """
+        return self.unit_host
+
+    @staticmethod
+    def _generate_relation_departed_key(rel_id: int) -> str:
+        """Generates the relation departed key for a specified relation id."""
+        return f"relation_{rel_id}_departed"
+
+    def open_mongos_port(self) -> None:
+        """Opens the mongos port for TCP connections."""
+        self.unit.open_port("tcp", Config.MONGOS_PORT)
+
+    def is_role(self, role_name: str) -> bool:
+        """Checks if application is running in provided role."""
+        return self.role == role_name
+
+    def is_db_service_ready(self) -> bool:
+        """Returns True if the underlying database service is ready."""
+        with MongosConnection(self.mongos_config) as mongos:
+            return mongos.is_ready
+
+    def _push_keyfile_to_workload(self, container: Container) -> None:
+        """Upload the keyFile to a workload container."""
+        keyfile = self.get_secret(APP_SCOPE, Config.Secrets.SECRET_KEYFILE_NAME)
+        if not keyfile:
+            raise MissingSecretError(f"No secret defined for {APP_SCOPE}, keyfile")
+        else:
+            self.push_file_to_unit(
+                container=container,
+                parent_dir=Config.MONGOD_CONF_DIR,
+                file_name=Config.TLS.KEY_FILE_NAME,
+                file_contents=keyfile,
+            )
+
+    @staticmethod
+    def _pull_licenses(container: Container) -> None:
+        """Pull licences from workload."""
+        licenses = [
+            "snap",
+            "rock",
+            "percona-server",
+        ]
+
+        for license_name in licenses:
+            try:
+                license_file = container.pull(
+                    path=Config.get_license_path(license_name)
+                )
+                f = open(f"LICENSE_{license_name}", "x")
+                f.write(str(license_file.read()))
+                f.close()
+            except FileExistsError:
+                pass
+
+    @staticmethod
+    def _set_data_dir_permissions(container: Container) -> None:
+        """Ensure the data directory for mongodb is writable for the "mongodb" user.
+
+        Until the ability to set fsGroup and fsGroupChangePolicy via Pod securityContext
+        is available, we fix permissions incorrectly with chown.
+        """
+        for path in [Config.DATA_DIR, Config.LOG_DIR, Config.LogRotate.LOG_STATUS_DIR]:
+            paths = container.list_files(path, itself=True)
+            assert (
+                len(paths) == 1
+            ), "list_files doesn't return only the directory itself"
```

Comment:
We should really avoid having `assert`s in the non-test code

Reply:
agreed - this was written by Dmitry and brought over here. I will update